src/datasets/arrow_dataset.py

```diff
@@ -159,6 +160,168 @@ def version(self):
         return self._info.version
 
 
+class TensorflowDatasetMixIn:
+    def __init__(self):
+        pass
+
+    @staticmethod
+    def _get_output_signature(dataset, test_batch, batch_size):
+        import tensorflow as tf
+
+        signatures = dict()
+        for column, col_feature in dataset.features.items():
+            if hasattr(col_feature, "feature"):
+                dtype_str = col_feature.feature.dtype
+            else:
+                dtype_str = col_feature.dtype
+            if dtype_str.startswith("int") or dtype_str.startswith("uint"):
+                dtype = tf.int64
+            elif dtype_str.startswith("float"):
+                dtype = tf.float32
+            else:
+                raise ValueError(f"Could not convert datatype {dtype_str} in column {column}!")
+
+            if isinstance(col_feature, (Value, ClassLabel)):
+                shape = [batch_size]
+            elif isinstance(col_feature, _ArrayXD):
+                shape = [batch_size] + list(col_feature.shape)
+            elif isinstance(col_feature, Sequence):
+                shape = [batch_size, col_feature.length]
+            else:
+                raise ValueError(
+                    f"Couldn't parse feature {column} with type {type(col_feature)}! "
+                    "This may indicate a column was included with an unusual datatype "
+                    "that we were unable to process correctly. "
+                    "If you're getting this error with one of our datasets, and you're "
+                    "sure the column should be convertable to tf.Tensor, please "
+                    "file an issue at github.com/huggingface/datasets and tag "
+                    "@rocketknight1!"
+                )
+            shape = [dim if dim != -1 else None for dim in shape]
+
+            signatures[column] = tf.TensorSpec(shape=shape, dtype=dtype)
+
+        # Catching columns added by the collate_fn, such as MLM labels
+        for column, tensor in test_batch.items():
+            if column in signatures:
+                continue
+            if column.startswith("label") and "input_ids" in signatures:
+                shape = signatures["input_ids"].shape
+            else:
+                # If this doesn't look like LM labels that got added by the collate_fn, let's not say anything
+                # about the dimensions we're unsure of
+                shape = [batch_size] + [None for dim in tensor.shape.as_list()[1:]]
+            signatures[column] = tf.TensorSpec(shape=shape, dtype=tensor.dtype)
+        return signatures
+
+    def to_tf_dataset(
+        self,
+        columns,
+        batch_size,
+        shuffle,
```

Comment:
Could they be optional parameters ?

Reply:
I was thinking about that! It's unclear to me what the defaults should be for `columns` or `batch_size` though, and I really wanted `shuffle` to be a required parameter to ensure people were aware of it, and that they didn't accidentally shuffle or skip shuffling their data when they didn't mean to.

I could maybe set `batch_size` to something like 32 by default and leave the other two as required parameters?

Reply:
Oh I see your point about shuffle.
And actually thinking more about it, it looks like we should require `batch_size` as well no ?

Maybe if columns is not specified then all of them are used ?

Reply:
I was thinking about that, but usually our datasets have one or more string columns that we don't want, so the default of using all columns will probably not work most of the time. It'd be nice if we had some way to auto-detect relevant columns, but I can't think of how we'd do that, so I think the safest thing to do is to just ask them to specify.